File: frontend-app/src/features/dashboard/api/metrics.ts
import { tenantApiRequest } from '@/config/api';

// Types for metrics responses
export interface OffersMetrics {
  draft: number;
  pending_review: number;
  approved: number;
  rejected: number;
  retired: number;
}

export interface CampaignsMetrics {
  draft: number;
  approved: number;
  active: number;
  paused: number;
  completed: number;
}

export interface CampaignCustomer {
  campaign_id: number;
  campaign_name: string;
  sent_campaigns_customers: number;
  accepted_campaigns: number;
  percentage_accepted: number;
}

export interface CampaignCustomersResponse {
  campaigns: CampaignCustomer[];
}

export interface DeliveryStatusResponse {
  pending: number;
  sent: number;
  declined: number;
  accepted: number;
}

export interface CustomerSegmentDistribution {
  segment: string;
  count: number;
  percentage: number;
}

export interface CustomerSegmentsResponse {
  total_customers: number;
  segments: CustomerSegmentDistribution[];
}

export interface DeliveryStatusBreakdown {
  pending: number;
  sent: number;
  declined: number;
  accepted: number;
}

export interface CampaignDeliveryStatus {
  campaign_id: number;
  campaign_name: string;
  delivery_status: DeliveryStatusBreakdown;
}

// API functions to fetch metrics
export const fetchDeliveryStatus = async (tenantName: string): Promise<DeliveryStatusResponse> => {
  return tenantApiRequest<DeliveryStatusResponse>(
    'GET',
    tenantName,
    '/metrics/delivery-status'
  );
};

export const fetchOffersMetrics = async (tenantName: string): Promise<OffersMetrics> => {
  return tenantApiRequest<OffersMetrics>(
    'GET',
    tenantName,
    '/metrics/offers-metrics'
  );
};

export const fetchCampaignsMetrics = async (tenantName: string): Promise<CampaignsMetrics> => {
  return tenantApiRequest<CampaignsMetrics>(
    'GET',
    tenantName,
    '/metrics/campaigns-metrics'
  );
};

export const fetchCampaignCustomers = async (tenantName: string): Promise<CampaignCustomersResponse> => {
  return tenantApiRequest<CampaignCustomersResponse>(
    'GET',
    tenantName,
    '/metrics/campaign-customers'
  );
};

export const fetchCustomerSegments = async (tenantName: string): Promise<CustomerSegmentsResponse> => {
  return tenantApiRequest<CustomerSegmentsResponse>(
    'GET',
    tenantName,
    '/metrics/customer-segments'
  );
};

export const fetchCampaignDeliveryStatus = async (tenantName: string): Promise<CampaignDeliveryStatus[]> => {
  return tenantApiRequest<CampaignDeliveryStatus[]>(
    'GET',
    tenantName,
    '/metrics/campaign-delivery-status'
  );
}; 

File: backend-apis/app/schemas/metrics.py
from typing import List, Dict, Optional
from pydantic import BaseModel


class ActiveCustomersResponse(BaseModel):
    count: int
    
    
class OffersMetricsResponse(BaseModel):
    draft: Optional[int] = 0
    pending_review: Optional[int] = 0
    approved: Optional[int] = 0
    rejected: Optional[int] = 0
    retired: Optional[int] = 0
    
    
class CampaignsMetricsResponse(BaseModel):
    draft: Optional[int] = 0
    approved: Optional[int] = 0
    active: Optional[int] = 0
    paused: Optional[int] = 0
    completed: Optional[int] = 0


class CampaignCustomerStats(BaseModel):
    campaign_id: int
    campaign_name: str
    sent_campaigns_customers: int
    accepted_campaigns: int
    percentage_accepted: float


class CampaignCustomersResponse(BaseModel):
    campaigns: List[CampaignCustomerStats]
    
    
# Additional metrics schemas
class CustomerSegmentDistribution(BaseModel):
    segment: str
    count: int
    percentage: float


class CustomerSegmentsResponse(BaseModel):
    total_customers: int
    segments: List[CustomerSegmentDistribution] 


class DeliveryStatusResponse(BaseModel):
    pending: int = 0
    sent: int = 0
    declined: int = 0
    accepted: int = 0


class CampaignDeliveryStatus(BaseModel):
    campaign_id: int
    campaign_name: str
    delivery_status: Dict[str, int]
